apiv2/db_import/importers/base_importer.py

```diff
@@ -0,0 +1,299 @@
+        query = session.scalars(
+            sa.select(klass).where(sa.and_(*[getattr(klass, k) == v for k, v in self.get_filters().items()])),
+        ).all()
+        all_records = []
+        for record in query:
+            all_records.append(record)
+            hashed_records[self.get_hash_value(record)] = record
+        return all_records, hashed_records
+
+    def update_data_map(self, data_map: dict[str, Any], metadata: dict[str, Any], index: int) -> dict[str, Any]:
+        """
+        Method to be overridden in cases where custom data mapping is necessary based on index value of the record in
+        the list.
+        """
+        return data_map
+
+    def import_to_db(self) -> Base:
+        """
+        Gets all the existing objects as a map for the current filter criteria. It creates/updates the records iterating
+        over the metadata list, and popping any existing value from the map. After the iteration, the map only contains
+        data that are no longer valid. Those stale records are deleted from the table.
+        """
+        klass = self.get_db_model_class()
+        data_map = self.get_data_map()
+        records_to_delete, hashed_records = self.get_existing_objects()
+        metadata_list = self.metadata if isinstance(self.metadata, list) else [self.metadata]
+        session = self.config.get_db_session()
+
+        db_obj = None
+        for index, entry in enumerate(metadata_list):
+            entry_data_map = self.update_data_map(data_map, entry, index)
+            hash_values = [str(map_to_value(id_field, entry_data_map, entry)) for id_field in self.get_id_fields()]
+            db_obj = hashed_records.get("-".join(hash_values))
+            if not db_obj:
+                db_obj = klass()
+            if db_obj in records_to_delete:
+                records_to_delete.remove(db_obj)
+
+            for db_key, _data_path in entry_data_map.items():
+                setattr(db_obj, db_key, map_to_value(db_key, entry_data_map, entry))
+            session.add(db_obj)
+
+        for stale_obj in records_to_delete:
+            logger.info("Deleting record of %s with id=%d and key=%s", klass, stale_obj.id, self.get_hash_value(stale_obj))
+            session.delete(stale_obj)
+        session.flush()
+        return db_obj
+
+
+class AuthorsStaleDeletionDBImporter(StaleDeletionDBImporter):
+    """
+    Supports insert of new authors, update of existing authors, and delete of stale authors. This class iterates over a
+    list of individual authors.
+    """
+
+    def update_data_map(self, data_map: dict[str, Any], metadata: dict[str, Any], index: int) -> dict[str, Any]:
+        """Adds author list order based on the index position to the author data"""
+        if metadata.get("author_list_order"):
+            return data_map
+        return {**data_map, **{"author_list_order": index + 1}}
+
+
+class StaleParentDeletionDBImporter(StaleDeletionDBImporter):
```

Comment:
I think this class might no longer be needed with the cascade delete in place.

Reply:
We still kind of do -- the interfaces are different for `StaleDeletionDBImporter` and `StaleParentDeletionDBImporter` - the first one expects to import all child rows at once, and the second expects to import one child row at a time. I'd rather do a bigger refactor of this code later if that's ok?